bench.c

```diff
@@ -0,0 +1,188 @@
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#define TARGET_IP "127.0.0.1"
+#define TARGET_PORT 12345
+#define BENCH_COUNT 10
+#define RESULT_FILE_NAME "bench.txt"
+
+/*
+ * TODO: length of the dummy msg should be vary since it's more conforms
+ * to reality
+ */
+const char *msg_dum = "dummy message";
+
+// vary-length message should shorter than this
+#define MAX_MSG_LEN 32
+
+/*
+ * Too much concurrent connection would be treat as sort of DDOS attack
+ * (mainly caused by configs (kernel: `tcp_max_syn_backlog` and
+ * `somaxconn`. fastecho: `DEFAULT_BACKLOG`). Nevertheless, default
+ * maximum number of fd per-process is limited under 1024, if you
+ * insist to proceed the benchmark, perform following modifications:
+ *       (PROCEDURE MAY VARY BETWEEN DIFFERENT DISTROS)
+ * For Ubuntu:
+ * (1)
+ * Use following commands to adjust kernel attributes:
+ *     a. `$ sudo sysctl net.core.somaxconn={depends on MAX_THREAD}`
+ *     b. `$ sudo sysctl net.ipv4.tcp_max_syn_backlog={ditto}`
+ * (`$ sudo sysctl net.core.somaxconn` can get current value. Note that
+ * `somaxconn` is max amount of established connection, whereas
+ * `tcp_max_syn_backlog` is max amount of connection at first step
+ * of 3-way handshake (SYN).)
+ *
+ * (2)
+ * Use command `$ ulimit -n {ditto}` to enlarge limitation of fd
+ * per-process. Note that this modification only effects on process
+ * which executes the command and its child processes.
+ *
+ * (3)
+ * Enlarge fastecho's `DEFAULT_BACKLOG`.
+ *
+ * Remember to reset the modifications after benchmarking to keep
+ * stability of running machine
+ */
+#define MAX_THREAD 1000
+
+pthread_t pt[MAX_THREAD];
+
+// block all worker before they are all ready to bombing kecho
+int rd_to_go = 0;
+
+pthread_mutex_t mutex_idx = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t mutex_wa_cond = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t cond_wait_all = PTHREAD_COND_INITIALIZER;
+
+long time_res[MAX_THREAD] = {0};
+int idx = 0;  // for indexing `time_res`
+FILE *bench_fd;
+
+static inline long time_diff_us(struct timeval *start, struct timeval *end)
+{
+    return ((end->tv_sec - start->tv_sec) * 1000000) +
+           (end->tv_usec - start->tv_usec);
+}
+
+static void *bomb_worker(void *arg)
+{
+    int sock_fd, rt;
+    char dummy[MAX_MSG_LEN];
+    struct timeval start, end;
+
+    // wait until all worker created
+    pthread_mutex_lock(&mutex_wa_cond);
+    while (!rd_to_go)
+        if ((rt = pthread_cond_wait(&cond_wait_all, &mutex_wa_cond))) {
+            puts("cond_wait failed!");
+            exit(-1);
+        }
+    pthread_mutex_unlock(&mutex_wa_cond);
+
+    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock_fd == -1) {
+        goto sock_init_fail;
+    }
+
+    struct sockaddr_in info;
+    bzero(&info, sizeof(info));
```

Comment:
Don't use bzero, which is deprecated. POSIX.1-2008 removes the specification of bzero().

Reply:
Instead, you can use C99 designated initializers as following:
```c
struct sockaddr_in info = {
    .sin_family = PF_INET,
    .sin_addr.s_addr = inet_addr(TARGET_IP),
    .sin_port = htons(TARGET_PORT),
};
```